components/eam/src/physics/crm/samomp/accelerate_crm.F90

```diff
@@ -0,0 +1,341 @@
+      !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+      !$omp target teams distribute parallel do collapse(2)
+      do k = 1, nzm
+        do icrm = 1, ncrms
+          ttend_acc(icrm,k) = tbaccel(icrm,k) - t0(icrm,k)
+          qtend_acc(icrm,k) = qtbaccel(icrm,k) - q0(icrm,k)
+          if (crm_accel_uv) then
+            utend_acc(icrm,k) = ubaccel(icrm,k) - u0(icrm,k)
+            vtend_acc(icrm,k) = vbaccel(icrm,k) - v0(icrm,k)
+          endif
+          if (abs(ttend_acc(icrm,k)) > ttend_threshold) then
+            ceaseflag = .true.
+          endif
+        enddo
+      enddo
+
+      !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+      !! Make sure it isn't insane
+      !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+      !$omp taskwait
+      if (ceaseflag) then ! special case for dT/dt too large
+        ! MSA will not be applied here or for the remainder of the CRM integration.
+        ! nstop must be updated to ensure the CRM integration duration is unchanged.
+        ! 
+        ! The effective MSA timestep is dt_a = crm_dt * (1 + crm_accel_factor). When
+        ! ceaseflag is triggered at nstep, we've taken (nstep - 1) previous steps of
+        ! size crm_dt * (1 + crm_accel_factor). The current step, and all future
+        ! steps, will revert to size crm_dt. Therefore, the total crm integration
+        ! time remaining after this step is
+        !     time_remaining = crm_run_time - (nstep - 1)* dt_a + crm_dt
+        !     nsteps_remaining = time_remaining / crm_dt
+        !     updated nstop = nstep + nsteps_remaining
+        ! Because we set nstop = crm_run_time / dt_a in crm_accel_nstop, subbing
+        ! crm_run_time = nstop * dt_a and working through algebra yields 
+        !     updated nstop = nstop + (nstop - nstep + 1) * crm_accel_factor.
+        write (*,*) 'accelerate_crm: mean-state acceleration not applied this step'
+        write (*,*) 'crm: nstop increased from ', nstop, ' to ', int(nstop+(nstop-nstep+1)*crm_accel_factor)
+        nstop = nstop + (nstop - nstep + 1)*crm_accel_factor ! only can happen once
+        return
+      endif
+
+      !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+      !! Apply the accelerated tendencies
+      !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+      !$omp target teams distribute parallel do collapse(4)
+      do k = 1, nzm
+        do j = 1, ny
+          do i = 1, nx
+            do icrm = 1, ncrms
+              ! don't let T go negative!
+              t(icrm,i,j,k) = max(tmin, t(icrm,i,j,k) + crm_accel_factor * ttend_acc(icrm,k))
+              if (crm_accel_uv) then
+                u(icrm,i,j,k) = u(icrm,i,j,k) + crm_accel_factor * utend_acc(icrm,k) 
+                v(icrm,i,j,k) = v(icrm,i,j,k) + crm_accel_factor * vtend_acc(icrm,k) 
+              endif
+              !$omp atomic update
```

Comment:
Why does this need to be atomic?

Reply:
I was a bit of confusing the atomic update in this one, it is used for identifying bugs. I will test to check whether we can remove it or not.